src/extension_commands.ml

```diff
@@ -84,10 +86,78 @@ let switch_impl_intf =
   in
   command Extension_consts.Commands.switch_impl_intf handler
 
-let register_all_commands extension instance =
-  let register_command { id; handler } =
-    let callback ~args:_ = handler instance () in
-    let disposable = Commands.registerCommand ~command:id ~callback in
-    ExtensionContext.subscribe extension ~disposable
+let remove_switch =
+  let handler (_ : Extension_instance.t) ~args =
+    let try_removing () =
+      let arg = List.hd_exn args in
+      let tree_item = TreeItem.t_of_js arg in
+      let dependency =
+        tree_item |> TreeItem.id |> Stdlib.Option.get
+        |> Treeview_switches.Dependency.of_string
+      in
+      match dependency with
+      | Dependency _ ->
+        Promise.return
+        @@ show_message `Warn
+             "Cannot delete a switch from a package dependency."
+      | Switch switch -> (
+        let open Promise.Syntax in
+        let* opam_opt = Opam.make () in
+        match opam_opt with
+        | None ->
+          Promise.return
+          @@ show_message `Warn "Opam could not be found on your system."
+        | Some opam -> (
+          let* result = Opam.remove_switch opam switch |> Cmd.output in
+          match result with
+          | Error err -> Promise.return @@ show_message `Error "%s" err
+          | Ok _ ->
+            let (_ : Ojs.t option Promise.t) =
+              Vscode.Commands.executeCommand
+                ~command:Extension_consts.Commands.refresh_switches ~args:[]
+            in
+            Promise.return
+            @@ show_message `Info "The switch has been removed successfully." )
+        )
+    in
+    let (_ : unit Promise.t) = try_removing () in
```

Comment:
Why do we need this `try_removing` wrapper? Also, why not wait for the promise to finish? (by returning it)

Reply:
> Why do we need this try_removing wrapper

We don't :)

> Also, why not wait for the promise to finish

The handler must have type `Extension_instance.t -> args:Ojs.t list -> unit`,  so I don't think we can wait for the promise.